pkg/provider/barbican/client.go

```diff
@@ -0,0 +1,236 @@
+			mapUUIDByUniqueNames[v.Name] = extractIdFromRef(v.SecretRef)
+		}
+	}
+
+	result := make(map[string][]byte, len(mapUUIDByUniqueNames))
+
+	for name, id := range mapUUIDByUniqueNames {
+		payload, err := secrets.GetPayload(c.client, id, secrets.GetPayloadOpts{PayloadContentType: "*/*"}).Extract()
+		metrics.ObserveAPICall(constants.ProviderBarbican, constants.CallBarbicanGetSecretPayload, err)
+		if err != nil {
+			return nil, err
+		}
+
+		result[name] = payload
+	}
+
+	return result, nil
+}
+
+// GetSecret returns a single secret from the provider.
+func (c *Client) GetSecret(ctx context.Context, ref esv1beta1.ExternalSecretDataRemoteRef) ([]byte, error) {
+	var (
+		name   = ref.Key
+		uuid   = getUUIDFromRemoteRef(name)
+		secret *secrets.Secret
+		err    error
+	)
+
+	metrics.ObserveAPICall(constants.ProviderBarbican, constants.CallBarbicanGetSecret, err)
+	if err != nil {
+		return nil, err
+	}
+
+	if uuid == "" {
+		secret, err = c.getByName(ctx, name)
+		if err != nil {
+			return nil, err
+		}
+		uuid = extractIdFromRef(secret.SecretRef)
+	}
+
+	return secrets.GetPayload(c.client, uuid, secrets.GetPayloadOpts{PayloadContentType: "*/*"}).Extract()
+}
+
+func (c *Client) Close(_ context.Context) error {
+	return nil
+}
+
+func (c *Client) Validate() (esv1beta1.ValidationResult, error) {
+	if c.storeKind == esv1beta1.ClusterSecretStoreKind && isReferentSpec(c.store) {
+		return esv1beta1.ValidationResultUnknown, nil
+	}
+	return esv1beta1.ValidationResultReady, nil
+}
+
+// GetSecretMap returns multiple k/v pairs from the provider.
+func (c *Client) GetSecretMap(ctx context.Context, ref esv1beta1.ExternalSecretDataRemoteRef) (map[string][]byte, error) {
+	data, err := c.GetSecret(ctx, ref)
+
+	var secretData map[string]interface{}
+	err = json.Unmarshal(data, &secretData)
+	if err != nil {
+		return nil, err
+	}
+	byteMap := make(map[string][]byte, len(secretData))
+	for k := range secretData {
+		byteMap[k], err = utils.GetByteValueFromMap(secretData, k)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return byteMap, nil
+}
+
+// PushSecret pushes a kubernetes secret key into barbican provider Secret.
+func (c *Client) PushSecret(ctx context.Context, secret *corev1.Secret, pushSecretData esv1beta1.PushSecretData) error {
+	payload := secret.Data[pushSecretData.GetSecretKey()]
+	name := pushSecretData.GetRemoteKey()
+
+	_, err := c.getByName(ctx, name)
+	if err == nil {
+		return nil
+	}
+
+	createOpts := secrets.CreateOpts{
+		Algorithm:          "aes",
+		BitLength:          256,
+		Mode:               "cbc",
+		Name:               name,
+		Payload:            string(payload),
+		PayloadContentType: "text/plain",
+		SecretType:         secrets.OpaqueSecret,
+	}
+
+	_, err = secrets.Create(c.client, createOpts).Extract()
+	metrics.ObserveAPICall(constants.ProviderBarbican, constants.CallBarbicanCreateSecret, err)
+	return err
+}
+
+func (s *Client) getByName(ctx context.Context, name string) (*secrets.Secret, error) {
```

Comment:
This is weird. Why are we `listing` by name? That implies that multiple secrets might have the same name and or at worst, multiple secrets will be displayed and we aren't a 100% sure that we got the actual secret that we wanted to.